webapp/src/ts/services/xml-forms.service.ts

```diff
@@ -164,7 +164,16 @@ export class XmlFormsService {
       }
       if (form.context.expression) {
         try {
-          return this.evaluateExpression(form.context.expression, options.doc, user, options.contactSummary);
+          const evaluatedExpression = this.evaluateExpression(
```

Comment:
I'm wondering if we can change this so we only ever call `evaluateExpression` and `authService.has` once per form. 
For example, extracting them into separate functions like: 
```
private checkFormExpression(form, doc, user, contactSummary) {
    if (!form.context.expression) {
      return true;
    }

    try {
      return this.evaluateExpression(
        form.context.expression,
        doc,
        user,
        contactSummary
      );
    } catch(err) {
      console.error(`Unable to evaluate expression for form: ${form._id}`, err);
      return false;
    }
  }

  private checkFormPermissions(form) {
    if (!form.context.permission) {
      return true;
    }

    return this.authService.has(form.context.permission);
  }
```
And then doing:  
```
return this
      .filterContactTypes(form.context, options.doc)
      .then(valid => valid && this.checkFormPermissions(form))
      .then(valid => valid && this.checkFormExpression(form, options.doc, user, options.contactSummary));
```

How does that sound?

Reply:
Brilliant! I have made the refactor as suggested.